IdentityCore/src/throttling/MSIDThrottlingService.m

```diff
@@ -0,0 +1,471 @@
+    if (!cacheRecord)
+    {
+        // we just log error (if any) and keep moving to the next UIRequired check
+        if (error)
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, self.context, BASE_MSG_READING_ERROR, error);
+        }
+        return NO;
+    }
+    else
+    {
+        [self updateServerTelemetry:cacheRecord];
+        resultBlock(YES, cacheRecord.cachedErrorResponse);
+        return YES;
+    }
+}
+
+/**
+ return cached error if a request is UIRequired in db.
+ If it's a hit, we also update telemetry.
+ */
+- (BOOL)isUIRequiredThrottleType:(id<MSIDThumbprintCalculatable> _Nonnull)request
+                    resultBlock:(nonnull MSIDThrottleResultBlock)resultBlock
+{
+    NSError *error = nil;
+    NSString *fullRequestThumbprint = [request fullRequestThumbprint];
+    MSIDThrottlingCacheRecord *cacheRecord = [self.cacheService objectForKey:fullRequestThumbprint                                                  error:&error];
+    if (!cacheRecord)
+    {
+        if (error)
+        {
+            MSID_LOG_WITH_CTX(MSIDLogLevelError, self.context, BASE_MSG_READING_ERROR, error);
+        }
+        return NO;
+    }
+    else
+    {
+        NSDate *currentTime = [NSDate date];
+        NSDate *lastRefreshTime = [MSIDThrottlingService getLastRefreshTimeWithContext:self.context error:&error];
+        // If currentTime is later than the expiration Time or the lastRefreshTime is later then the expiration Time, we clear the cache record
+        if ([currentTime compare:cacheRecord.expirationTime] != NSOrderedAscending
+            || (lastRefreshTime && [lastRefreshTime compare:cacheRecord.expirationTime] != NSOrderedAscending))
+        {
+            [self.cacheService removeObjectForKey:fullRequestThumbprint error:&error];
+            if (error)
+            {
+                MSID_LOG_WITH_CTX(MSIDLogLevelError, self.context, BASE_MSG_READING_ERROR, error);
+            }
+            return NO;
+        }
+        
+        // Expiration time still valid, we have a valid throttle case here so return cache response to calling
+        [self updateServerTelemetry:cacheRecord];
+        resultBlock(YES, cacheRecord.cachedErrorResponse);
+        return YES;
+    }
+}
+
+/**
+ Prepare record and update to throttling cache
+ */
+- (void)createDBRecordAndUpdateWithRequest:(id<MSIDThumbprintCalculatable> _Nonnull)request
+                             errorResponse:(NSError * _Nullable)errorResponse
+                           isSSOExtRequest:(BOOL)isSSOExtRequest
+                              throttleType:(MSIDThrottlingType)throttleType
+                               returnError:(NSError *_Nullable *_Nullable)error
+{
+    NSString *thumbprint = nil;
+    NSString *throttleTypeString = nil;
+    NSInteger throttleDuration = 0;
+    
+    NSDate *retryHeaderDate = [self getRetryDateFromErrorResponse:errorResponse];
+    
+    switch (throttleType)
+    {
+        case MSIDThrottlingType429:
```

Comment:
I see a lot of checks for throttle type in methods here. Should we maybe create classes instead with a common interface and this class be responsible for common logic?

Reply:
Addressed in the latest commit, please help to review.